File: src/main/scala/workshop/part3/SuperComputeActor.scala
package workshop.part3

import scala.language.postfixOps
import akka.actor.{OneForOneStrategy, ActorRef, Props, Actor}
import workshop.work.RiskyWork
import akka.actor.SupervisorStrategy.Stop
import akka.event.Logging
import scala.concurrent.duration._

class SuperComputeActor() extends Actor {
  val log = Logging(context.system, this)

  override val supervisorStrategy =
      OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1 minute, loggingEnabled = false) {
        case _ => {
          log.info("Stopping problem actor!")
          Stop
        }
      }

  def receive = {
    case riskyWork : RiskyWork => {
      val worker: ActorRef = context.actorOf(Props[Worker])
      worker.tell(riskyWork, sender())
    }
  }
}

class Worker extends Actor {
  def receive = {
    case riskyWork: RiskyWork => {
      sender ! riskyWork.perform()
      context.stop(self)
    }
  }
}

File: src/main/java/workshop/part3/SuperComputeActor.java
package workshop.part3;

import akka.actor.*;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;
import scala.PartialFunction;
import scala.concurrent.duration.Duration;
import scala.runtime.BoxedUnit;
import workshop.work.Work;

import static akka.actor.SupervisorStrategy.stop;
import static java.util.concurrent.TimeUnit.MINUTES;

public class SuperComputeActor extends AbstractActor {

    private final LoggingAdapter log = Logging.getLogger(context().system(), this);

    @Override
    public PartialFunction<Object, BoxedUnit> receive() {
        //TODO
        return ReceiveBuilder.matchAny(x -> {}).build();
    }
}